README.md

## Overview

Erlang Modebus Client

## Usage

### Connect

```
{ok, Device} = emodbus:connect("localhost", 502).
```

### Read Coils

```
emodbus:read_coils(Device, Offset, Count).
```

### Read HRegs

```
emodbus:read_hregs(Device, Offset, Count).
```

### Write Coil

```
%% Coil = 0 | 1
emodbus:write_coil(Device, Offset, Coil).
```

### Write HReg

```
emodbus:write_hreg(Device, Offset, Value).
```

### Disconnect

```
emodubs:disconnect(Device).

```

## License

The MIT License

## Author

feng at emqtt.io



CHANGELOG.md

eModbus ChangeLog
==================

0.3-alpha (2015/06/02)
------------------------

Test with modbus server


0.2-alpha (2015/06/01)
------------------------

Simplify the design, only emodbus.erl left.


0.1-alpha (2015/05/13)
------------------------

first release




test/emodbus_test.erl
%%%-----------------------------------------------------------------------------
%%% Copyright (c) 2015 Feng Lee <feng@emqtt.io>, All Rights Reserved.
%%%
%%% Permission is hereby granted, free of charge, to any person obtaining a copy
%%% of this software and associated documentation files (the "Software"), to deal
%%% in the Software without restriction, including without limitation the rights
%%% to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
%%% copies of the Software, and to permit persons to whom the Software is
%%% furnished to do so, subject to the following conditions:
%%%
%%% The above copyright notice and this permission notice shall be included in all
%%% copies or substantial portions of the Software.
%%%
%%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
%%% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
%%% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
%%% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
%%% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
%%% OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
%%% SOFTWARE.
%%%-----------------------------------------------------------------------------
-module(emodbus_test).

-ifdef(TEST).

-include_lib("eunit/include/eunit.hrl").


-endif.



test/emodbus_frame_test.erl
%%%-----------------------------------------------------------------------------
%%% Copyright (c) 2015 Feng Lee <feng@emqtt.io>, All Rights Reserved.
%%%
%%% Permission is hereby granted, free of charge, to any person obtaining a copy
%%% of this software and associated documentation files (the "Software"), to deal
%%% in the Software without restriction, including without limitation the rights
%%% to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
%%% copies of the Software, and to permit persons to whom the Software is
%%% furnished to do so, subject to the following conditions:
%%%
%%% The above copyright notice and this permission notice shall be included in all
%%% copies or substantial portions of the Software.
%%%
%%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
%%% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
%%% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
%%% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
%%% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
%%% OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
%%% SOFTWARE.
%%%-----------------------------------------------------------------------------
-module(emodbus_frame_test).

-ifdef(TEST).

-include("emodbus.hrl").

-include_lib("eunit/include/eunit.hrl").

-import(emodbus_frame, [init/0, parse/2]).

parse_test() ->

    parse(<<0,1,0,0,0,6,1,1,0,10,0,10>>, init(server)),

    ?assertMatch({more, _Cont}, parse(<<01, 16#5B, 00, 00, 00>>, init())),

    ?assertMatch({ok, #modbus_frame{hdr = #mbap_header{tid = 16#15B, length = 6},
                                    pdu = #modbus_pdu{opcode = 01}}, <<>>}, 
                 begin
                     {more, Cont} = parse(<<01, 16#5B, 00, 00, 00>>, init()),
                     parse(<<06, 01, 01, 00, 00, 00, 16#0A>>, Cont)
                 end),
    
    ?assertMatch({ok, #modbus_frame{hdr = #mbap_header{tid = 16#15B, length = 6},
                                    pdu = #modbus_pdu{opcode = 01}}, <<>>},
                parse(<<01, 16#5B, 00, 00, 00, 06, 01, 01, 00, 00, 00, 16#0A>>, init())),
    ?assertMatch({ok, #modbus_frame{hdr = #mbap_header{tid = 16#15B, length = 5},
                                    pdu = #modbus_pdu{opcode = 01}}, <<>>},
                parse(<<01, 16#5B, 00, 00, 00, 05, 01, 01, 02, 16#7F, 00>>, init())),
    ?assertMatch({ok, #modbus_frame{hdr = #mbap_header{tid = 16#34, length = 6},
                                    pdu = #modbus_pdu{opcode = 3}}, <<>>},
                parse(<<00, 16#34, 00, 00, 00, 06, 01, 03, 00, 00, 00, 16#0A>>, init())),

    ?assertMatch({ok, #modbus_frame{hdr = #mbap_header{tid = 16#34, length = 17},
                                    pdu = #modbus_pdu{opcode = 3}}, <<0,0,0,0,0,0>>},
                parse(<<00, 16#34, 00, 00, 00, 17, 01, 03, 14, 00, 01, 00, 02, 00, 03, 00, 04, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00>>, init())).

-endif.


    %?assertMatch({ok, #modbus_frame{hdr = #mbap_header{tid = 16#34, length = 17},
    %                                pdu = #modbus_pdu{opcode = 3}}, <<>>},
    %             begin
    %                 {more, Cont} = parse(<<00, 16#34, 00, 00, 00, 17, 01, 03, 14, 00, 01>>, init()),
    %                 parse(<<00, 02, 00, 03, 00, 04, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00>>, Cont)
    %             end),


src/emodbus.erl
%%%-----------------------------------------------------------------------------
%%% Copyright (c) 2015 Feng Lee <feng@emqtt.io>, All Rights Reserved.
%%%
%%% Permission is hereby granted, free of charge, to any person obtaining a copy
%%% of this software and associated documentation files (the "Software"), to deal
%%% in the Software without restriction, including without limitation the rights
%%% to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
%%% copies of the Software, and to permit persons to whom the Software is
%%% furnished to do so, subject to the following conditions:
%%%
%%% The above copyright notice and this permission notice shall be included in all
%%% copies or substantial portions of the Software.
%%%
%%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
%%% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
%%% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
%%% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
%%% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
%%% OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
%%% SOFTWARE.
%%%-----------------------------------------------------------------------------
%%% @doc
%%% emodbus TCP client.
%%%
%%% @end
%%%-----------------------------------------------------------------------------
-module(emodbus).

-include("emodbus.hrl").

%% API Exports
-export([connect/0, connect/2, connect/3,
         disconnect/1]).

%% modbus API Exports
-export([read_coils/3,
         read_inputs/3,
         read_hregs/3,
         read_iregs/3,
         write_coil/3,
         write_hreg/3,
         write_coils/3,
         write_hregs/3]).

-behaviour(gen_server).

%% gen_server Function Exports
-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
         terminate/2, code_change/3]).

-record(state, {
        socket      :: inet:socket(),
        tid = 1     :: 1..16#ff,
        unit_id = 1 :: byte()
 }).

-define(MBAP_LENGTH, 7).

-define(CALL_TIMEOUT, 60000).

-define(SOCK_TIMEOUT, 30000).

-define(RECV_TIMEOUT, 10000).

%% Modbus Socket Options
-define(SOCKOPTS, [
        binary,
        {active, false},
        {packet, 0},
        {reuseaddr, true},
        {nodelay,   true}
]).

connect() ->
    connect("localhost", ?MODBUS_PORT).

connect(Host, Port) ->
    connect(Host, Port, 1).

connect(Host, Port, UnitId) ->
    gen_server:start_link(?MODULE, [Host, Port, UnitId], []).

disconnect(Device) ->
    call(Device, stop).

read_coils(Device, Offset, Count) ->
    call(Device, #modbus_req{funcode = ?FUN_CODE_WRITE_COILS, offset = Offset, count = Count}).

read_inputs(Device, Offset, Count) ->
    call(Device, #modbus_req{funcode = ?FUN_CODE_READ_INPUTS, offset = Offset, count = Count}).

read_hregs(Device, Offset, Count) ->
    call(Device, #modbus_req{funcode = ?FUN_CODE_READ_HREGS, offset = Offset, count = Count}).

read_iregs(Device, Offset, Count) ->
    call(Device, #modbus_req{funcode = ?FUN_CODE_READ_IREGS, offset = Offset, count = Count}).

write_coil(Device, Offset, Coil) when Coil =:= 0 orelse Coil =:= 1 ->
    call(Device, #modbus_req{funcode = ?FUN_CODE_WRITE_COIL, offset = Offset, count = onoff(Coil)}).

write_hreg(Device, Offset, Value) when Value >= 0 andalso Value =< 16#FFFF ->
    call(Device, #modbus_req{funcode = ?FUN_CODE_WRITE_HREG, offset = Offset, count = Value}).

write_coils(_Device, _Offset, _Coils) ->
    {error, unsupport}.

write_hregs(_Device, _Offset, _Value) ->
    {error, unsupport}.

call(Device, Request) ->
    gen_server:call(Device, Request, ?CALL_TIMEOUT).

init([Host, Port, UnitId]) ->
    case gen_tcp:connect(Host, Port, ?SOCKOPTS, ?SOCK_TIMEOUT) of
        {ok, Sock} -> 
            {ok, #state{socket = Sock, unit_id = UnitId}};
        {error, Error} -> 
            {stop, {error, Error}}
    end.

handle_call(Req, _From, State) when is_record(Req, modbus_req) ->
    case catch request(Req, State) of
        {'EXIT', Error} ->
            {reply, {error, Error}, next_id(State)};
        Response ->
            {reply, Response, next_id(State)}
    end;

handle_call(stop, _From, State=#state{socket = Sock}) ->
	gen_tcp:close(Sock),
	{stop, normal, stopped, State};

handle_call(_Req, _From, State) ->
    {reply, {error, badreq}, State}.

handle_cast(_Msg, State) ->
    {noreply, State}.

handle_info(_Info, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

request(Req, State) ->
    send_and_recv(Req, State).

send_and_recv(Req, State) ->
    case send(Req, State) of
        ok ->
            case recv(State) of
                {ok, Frame} ->
                    response(Req, Frame);
                {error, Reason} ->
                    {error, Reason}
            end;
        {error, Reason} ->
            {error, Reason}
    end.

response(#modbus_req{funcode = FunCode, count = Count}, #modbus_frame{funcode = FunCode, payload = Payload}) ->
    case FunCode of
        ?FUN_CODE_READ_COILS -> 
            <<_:8, Data/binary>> = Payload,
            {ok, bits(Data, Count, [])};
        ?FUN_CODE_READ_INPUTS ->
            <<_:8, Data/binary>> = Payload,
            {ok, bits(Data, Count, [])};
        ?FUN_CODE_READ_HREGS ->
            <<_:8, Data/binary>> = Payload,
            {ok, words(Data)};
        ?FUN_CODE_READ_IREGS ->
            <<_:8, Data/binary>> = Payload,
            {ok, words(Data)};
        ?FUN_CODE_WRITE_COIL ->
            <<_:16, Coil:16>> = Payload,
            {ok, onoff(Coil)};
        ?FUN_CODE_WRITE_HREG ->
            <<_:16, Val:16>> = Payload,
            {ok, Val}
    end;
    
response(#modbus_req{funcode = FunCode}, #modbus_frame{funcode = ErroCode, payload = Payload}) 
    when ErroCode == (FunCode + 16#80) -> {error, err(Payload)}.

send(#modbus_req{funcode = FunCode, offset = Offset, count = Count}, #state{tid = Tid, unit_id = UnitId, socket = Sock}) ->
    Hdr= #mbap_header{tid = Tid, unit_id = UnitId},
    Frame = #modbus_frame{header = Hdr, funcode = FunCode, payload = <<Offset:16, Count:16>>},
    gen_tcp:send(Sock, serialize(Frame)).

serialize(#modbus_frame{header = #mbap_header{tid = Tid, unit_id = Uid},
                        funcode = FunCode, payload = Payload}) ->
    Len = 2 + size(Payload),
    <<Tid:16, 0:16, Len:16, Uid:8, FunCode:8, Payload/binary>>.
    
recv(State) ->
    case recv(header, State) of
        {ok, Header} ->
            recv(payload, Header, State);
        {error,Error} ->
            {error,Error}
    end.

recv(header, State = #state{tid = Tid, socket = Sock}) ->
    case gen_tcp:recv(Sock, ?MBAP_LENGTH, ?RECV_TIMEOUT) of
        {ok, <<Tid:16, 0:16, Len:16, Uid:8>>} ->
            recv(payload, #mbap_header{tid = Tid, length = Len, unit_id = Uid}, State);
        {ok, Header} ->
            error_logger:error_msg("Response cannot match request: request tid=~p, response header =~p", [Tid, Header]),
            {error, badresp};
        {error, Reason} ->
            {error, Reason}
    end.

recv(payload, Header = #mbap_header{length = Len}, #state{socket = Sock}) ->
    case gen_tcp:recv(Sock, Len - 1, ?RECV_TIMEOUT) of
        {ok, <<FunCodeOrErr:8, Payload/binary>>} ->
            {ok, #modbus_frame{header = Header, funcode = FunCodeOrErr, payload = Payload}};
        {error, Reason} ->
            {error, Reason}
    end.

next_id(State = #state{tid = Tid}) when Tid >= 16#EFFF ->
    State#state{tid = 1};

next_id(State = #state{tid = Tid}) ->
    State#state{tid = Tid+1}.

onoff(1) -> 16#FF00;
onoff(16#FF00) -> 1;
onoff(_) -> 0.

bits(_Data, 0, Acc) ->
    lists:reverse(Acc);
bits(Data, Count, Acc) ->
    <<Bit:1, Left/binary>> = Data,
    bits(Left, Count-1, [Bit|Acc]).

words(Data) ->
    words(Data, []).
words(<<>>, Acc) ->
    lists:reverse(Acc);
words(<<I:16, Left/binary>>, Acc) ->
    words(Left, [I|Acc]).

err(<<Code:8>>) -> err(Code); 
err(1)  -> illegal_function;
err(2)  -> illegal_data_address;
err(3)  -> illegal_data_value;
err(4)  -> slave_device_failure;
err(5)  -> acknowledge;
err(6)  -> slave_device_busy;
err(8)  -> memory_parity_error;
err(10) -> gateway_path_unavailable;
err(11) -> gateway_target_device_failed_to_respond;
err(_)  -> unknown_response_code.



src/emodbus_app.erl
%%%-----------------------------------------------------------------------------
%%% Copyright (c) 2015 Feng Lee <feng@emqtt.io>, All Rights Reserved.
%%%
%%% Permission is hereby granted, free of charge, to any person obtaining a copy
%%% of this software and associated documentation files (the "Software"), to deal
%%% in the Software without restriction, including without limitation the rights
%%% to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
%%% copies of the Software, and to permit persons to whom the Software is
%%% furnished to do so, subject to the following conditions:
%%%
%%% The above copyright notice and this permission notice shall be included in all
%%% copies or substantial portions of the Software.
%%%
%%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
%%% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
%%% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
%%% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
%%% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
%%% OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
%%% SOFTWARE.
%%%-----------------------------------------------------------------------------
%%% @doc
%%% emodbus app.
%%%
%%% @end
%%%-----------------------------------------------------------------------------
-module(emodbus_app).

-behaviour(application).

%% Application callbacks
-export([start/2, stop/1]).

%%%=============================================================================
%%% Application callbacks
%%%=============================================================================

start(_StartType, _StartArgs) ->
    {ok, Sup} = emodbus_sup:start_link(),
    {ok, Enable} = application:get_env(server),
    start_server(Enable),
    {ok, Sup}.

start_server(enable) ->
    {ok, Listeners} = application:get_env(listeners),
    emodbus:open_listeners(Listeners);

start_server(_) ->
    ignore.

stop(_State) ->
    ok.



src/emodbus_crc16.erl
-module(emodbus_crc16).

-export([calc/1]).

-define(CRC16Def,[16#0000, 16#C0C1, 16#C181, 16#0140, 16#C301,  
16#03C0, 16#0280, 16#C241,
                    16#C601, 16#06C0, 16#0780, 16#C741, 16#0500,  
16#C5C1, 16#C481, 16#0440,
                    16#CC01, 16#0CC0, 16#0D80, 16#CD41, 16#0F00,  
16#CFC1, 16#CE81, 16#0E40,
                    16#0A00, 16#CAC1, 16#CB81, 16#0B40, 16#C901,  
16#09C0, 16#0880, 16#C841,
                    16#D801, 16#18C0, 16#1980, 16#D941, 16#1B00,  
16#DBC1, 16#DA81, 16#1A40,
                    16#1E00, 16#DEC1, 16#DF81, 16#1F40, 16#DD01,  
16#1DC0, 16#1C80, 16#DC41,
                    16#1400, 16#D4C1, 16#D581, 16#1540, 16#D701,  
16#17C0, 16#1680, 16#D641,
                    16#D201, 16#12C0, 16#1380, 16#D341, 16#1100,  
16#D1C1, 16#D081, 16#1040,
                    16#F001, 16#30C0, 16#3180, 16#F141, 16#3300,  
16#F3C1, 16#F281, 16#3240,
                    16#3600, 16#F6C1, 16#F781, 16#3740, 16#F501,  
16#35C0, 16#3480, 16#F441,
                    16#3C00, 16#FCC1, 16#FD81, 16#3D40, 16#FF01,  
16#3FC0, 16#3E80, 16#FE41,
                    16#FA01, 16#3AC0, 16#3B80, 16#FB41, 16#3900,  
16#F9C1, 16#F881, 16#3840,
                    16#2800, 16#E8C1, 16#E981, 16#2940, 16#EB01,  
16#2BC0, 16#2A80, 16#EA41,
                    16#EE01, 16#2EC0, 16#2F80, 16#EF41, 16#2D00,  
16#EDC1, 16#EC81, 16#2C40,
                    16#E401, 16#24C0, 16#2580, 16#E541, 16#2700,  
16#E7C1, 16#E681, 16#2640,
                    16#2200, 16#E2C1, 16#E381, 16#2340, 16#E101,  
16#21C0, 16#2080, 16#E041,
                    16#A001, 16#60C0, 16#6180, 16#A141, 16#6300,  
16#A3C1, 16#A281, 16#6240,
                    16#6600, 16#A6C1, 16#A781, 16#6740, 16#A501,  
16#65C0, 16#6480, 16#A441,
                    16#6C00, 16#ACC1, 16#AD81, 16#6D40, 16#AF01,  
16#6FC0, 16#6E80, 16#AE41,
                    16#AA01, 16#6AC0, 16#6B80, 16#AB41, 16#6900,  
16#A9C1, 16#A881, 16#6840,
                    16#7800, 16#B8C1, 16#B981, 16#7940, 16#BB01,  
16#7BC0, 16#7A80, 16#BA41,
                    16#BE01, 16#7EC0, 16#7F80, 16#BF41, 16#7D00,  
16#BDC1, 16#BC81, 16#7C40,
                    16#B401, 16#74C0, 16#7580, 16#B541, 16#7700,  
16#B7C1, 16#B681, 16#7640,
                    16#7200, 16#B2C1, 16#B381, 16#7340, 16#B101,  
16#71C0, 16#7080, 16#B041,
                    16#5000, 16#90C1, 16#9181, 16#5140, 16#9301,  
16#53C0, 16#5280, 16#9241,
                    16#9601, 16#56C0, 16#5780, 16#9741, 16#5500,  
16#95C1, 16#9481, 16#5440,
                    16#9C01, 16#5CC0, 16#5D80, 16#9D41, 16#5F00,  
16#9FC1, 16#9E81, 16#5E40,
                    16#5A00, 16#9AC1, 16#9B81, 16#5B40, 16#9901,  
16#59C0, 16#5880, 16#9841,
                    16#8801, 16#48C0, 16#4980, 16#8941, 16#4B00,  
16#8BC1, 16#8A81, 16#4A40,
                    16#4E00, 16#8EC1, 16#8F81, 16#4F40, 16#8D01,  
16#4DC0, 16#4C80, 16#8C41,
                    16#4400, 16#84C1, 16#8581, 16#4540, 16#8701,  
16#47C0, 16#4680, 16#8641,
                    16#8201, 16#42C0, 16#4380, 16#8341, 16#4100,  
16#81C1, 16#8081, 16#4040]).

calc(List) ->
   calc(List,16#FFFF).
calc(<<>>,CRC) ->
   CRC;
calc([],CRC) ->
   CRC;
calc(<<Value:8,Rest/binary>>,CRC) when Value =< 255->
   Index = (CRC bxor Value) band 255,
   NewCRC = (CRC bsr 8) bxor crc_index(Index),
   calc(Rest,NewCRC);
calc([Value|Rest],CRC) when Value =< 255->
   Index = (CRC bxor Value) band 255,
   NewCRC = (CRC bsr 8) bxor crc_index(Index),
   calc(Rest,NewCRC).

crc_index(N) ->
   lists:nth(N+1,?CRC16Def).



src/emodbus_device.erl
%%%-----------------------------------------------------------------------------
%%% Copyright (c) 2015 Feng Lee <feng@emqtt.io>, All Rights Reserved.
%%%
%%% Permission is hereby granted, free of charge, to any person obtaining a copy
%%% of this software and associated documentation files (the "Software"), to deal
%%% in the Software without restriction, including without limitation the rights
%%% to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
%%% copies of the Software, and to permit persons to whom the Software is
%%% furnished to do so, subject to the following conditions:
%%%
%%% The above copyright notice and this permission notice shall be included in all
%%% copies or substantial portions of the Software.
%%%
%%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
%%% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
%%% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
%%% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
%%% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
%%% OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
%%% SOFTWARE.
%%%-----------------------------------------------------------------------------
%%% @doc
%%% emodbus device module
%%%
%%% @end
%%%-----------------------------------------------------------------------------

-module(emodbus_device).

-include("emodbus.hrl").

-behaviour(gen_server).

%% API Exports
-export([start_link/0, start_link/1, stop/1]).

-export([read_coils/3,
         read_inputs/3,
         read_hregs/3,
         read_iregs/3,
         write_coil/3,
         write_coils/3,
         write_hregs/3]).

%% gen_server Function Exports
-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
         terminate/2, code_change/3]).

-type option() :: {host, inet:ip_address() | string()}
                | {port, inet:port_number()}
                | {timeout, pos_integer()}
                | {logger, atom() | {atom(), atom()}}.

-record(state, {name                :: atom(),
                host = "localhost"  :: inet:ip_address() | string(),
                port = ?MODBUS_PORT :: inet:port_number(),
                socket              :: inet:socket(),
                timeout             :: pos_integer(),
                pending_reqs,                
                tid = 1             :: 1..16#ff,
                unit_id = 1         :: byte(),
                parse_state,
                logger              :: gen_logger:logmod()}).

-define(TIMEOUT, 10000).

start_link() ->
    start_link([]).

-spec start_link([option()]) -> {ok, pid()} | {error, any()}.
start_link(Opts) ->
    gen_server:start_link(?MODULE, [Opts], []).

read_coils(Device, Offset, Count) ->
    call(Device, {read, coils, Offset, Count}).

read_inputs(Device, Offset, Count) ->
    call(Device, {read, inputs, Offset, Count}).

read_iregs(Device, Offset, Count) ->
    call(Device, {read, iregs, Offset, Count}).

read_hregs(Device, Offset, Count) ->
    call(Device, {read, hregs, Offset, Count}).

-spec write_coil(pid(), 0..16#ffff, on | off) -> {ok, on | off}.
write_coil(Device, Offset, Coil) when Coil =:= on; Coil =:= off ->
    call(Device, {write, coil, Offset, Coil}).

write_coils(Device, Offset, Coils) ->
    call(Device, {write, coils, Offset, Coils}).

write_hregs(Device, Offset, Value) ->
    call(Device, {write, hregs, Offset, Value}).

stop(Device) ->
    gen_server:call(Device, stop).

call(Device, Request) ->
    gen_server:call(Device, Request).

init([Opts]) ->
    Logger = gen_logger:new(proplists:get_value(logger, Opts, {console, debug})),

    Opts1 = proplists:delete(logger, Opts),

    State = init(Opts1, #state{host         = "localhost",
                               port         = ?MODBUS_PORT,
                               timeout      = ?TIMEOUT,
                               logger       = Logger,
                               unit_id      = 1,
                               pending_reqs = gb_trees:empty(),
                               parse_state  = emodbus_frame:init(client)}),

    case connect(State) of
        {ok, Sock}     -> {ok, run_sock(State#state{socket = Sock})};
        {error, Error} -> {stop, {error, Error}}
    end.

init([], State) ->
    State;
init([{host, Host} | Opts], State) ->
    init(Opts, State#state{host = Host});
init([{port, Port} | Opts], State) ->
    init(Opts, State#state{port = Port});
init([{timeout, Timeout} | Opts], State) ->
    init(Opts, State#state{timeout = Timeout * 1000});
init([{logger, Cfg} | Opts], State) ->
    init(Opts, State#state{logger = gen_logger:new(Cfg)});
init([{unit_id, Id} | Opts], State) ->
    init(Opts, State#state{unit_id = Id});
init([_Opt | Opts], State) ->
    init(Opts, State).

handle_call(Req = {read, coils, Offset, Count}, From, State) ->
    send(#modbus_req{funcode = ?FUN_CODE_READ_COILS,
                     offset  = Offset, quantity = Count}, State),
    {noreply, next_id(store_req(Req, From, State))};

handle_call(Req = {read, inputs, Offset, Count}, From, State) ->
    send(#modbus_req{funcode = ?FUN_CODE_READ_INPUTS,
                     offset  = Offset, quantity = Count}, State),
    {noreply, next_id(store_req(Req, From, State))};

handle_call(Req = {read, hregs, Offset, Count}, From, State) ->
    send(#modbus_req{funcode = ?FUN_CODE_READ_HREGS,
                     offset  = Offset, quantity = Count}, State),
    {noreply, next_id(store_req(Req, From, State))};

handle_call(Req = {write, coil, Offset, Coil}, From, State) ->
    send(#modbus_req{funcode = ?FUN_CODE_WRITE_COIL,
                     offset  = Offset, value = Coil}, State),
    {noreply, next_id(store_req(Req, From, State))};

handle_call(Req = {write, hreg, Offset, Val}, From, State) ->
    send(#modbus_req{funcode = ?FUN_CODE_WRITE_HREG,
                     offset = Offset, value = Val}, State),
    {noreply, next_id(store_req(Req, From, State))};

handle_call(Req = {write, coils, Offset, Coils}, From, State) ->
    send(#modbus_req{funcode = ?FUN_CODE_WRITE_COILS,
                     offset = Offset, value = Coils}, State),
    {noreply, next_id(store_req(Req, From, State))};

handle_call(Req = {write, hregs, Offset, Val}, From, State) ->
    send(#modbus_req{funcode = ?FUN_CODE_WRITE_HREGS,
                     offset = Offset, value = Val}, State),
    {noreply, next_id(store_req(Req, From, State))};
    
handle_call(stop, _From, State=#state{socket = Sock}) ->
	gen_tcp:close(Sock),
	{stop, normal, stopped, State};

handle_call(_Request, _From, State) ->
    {reply, ok, State}.

handle_cast(_Msg, State) ->
    {noreply, State}.

handle_info({tcp, Sock, Data}, State=#state{socket = Sock}) ->
    received(Data, run_sock(State));

handle_info({tcp_error, Sock, Reason}, State=#state{socket = Sock}) ->
    {stop, {shutdown, {tcp_error, Reason}}, State};

handle_info({tcp_closed, Sock}, State=#state{socket =Sock}) ->
	{stop, {shutdown, tcp_closed}, State};

handle_info(_Info, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

connect(#state{host = Host, port = Port, timeout = Timeout}) ->
    case gen_tcp:connect(Host, Port, ?MODBUS_SOCKOPTS, Timeout) of
        {ok, Sock}     -> {ok, Sock}; 
        {error, Error} -> {error, Error}
    end.

send(PDU, #state{tid = Tid, unit_id = Uid, socket = Sock}) ->
    Frame = #modbus_frame{hdr = #mbap_header{tid = Tid, unit_id = Uid}, pdu = PDU},
    gen_tcp:send(Sock, emodbus_frame:serialize(Frame)).

received(<<>>, State) ->
    {noreply, State, hibernate};

received(Bytes, State = #state{parse_state = ParseState}) ->
    case emodbus_frame:parse(Bytes, ParseState) of
        {more, ParseState1} ->
            {noreply, run_sock(State#state{parse_state = ParseState1}), hibernate};
        {ok, Frame, Rest} ->
            {ok, NewState} = handle_frame(Frame, State),
            received(Rest, NewState#state{parse_state = emodbus_frame:init(client)});
        {error, Error} ->
            {stop, {shutdown, Error}, State}
    end.

handle_frame(Frame = #modbus_frame{hdr = #mbap_header{tid = Tid},
                                   pdu = Resp}, State = #state{pending_reqs = Reqs}) ->
    io:format("RECV Frame: ~p~n", [Frame]),
    case gb_trees:lookup(Tid, Reqs) of
        {value, {From, _Req}} ->
            gen_server:reply(From, {ok, Resp#modbus_resp.bytes}),
            {ok, State#state{pending_reqs = gb_trees:delete(Tid, Reqs)}};
        none ->
            {ok, State}
    end.

run_sock(State=#state{socket = Sock}) ->
    inet:setopts(Sock, [{active, once}]), State.

next_id(State = #state{tid = Tid}) when Tid >= 16#EFFF ->
    State#state{tid = 1};

next_id(State = #state{tid = Tid}) ->
    State#state{tid = Tid+1}.

store_req(Req, From, State=#state{tid = Tid, pending_reqs = Reqs}) ->
    Reqs1 = gb_trees:enter(Tid, {From, Req}, Reqs),
    State#state{pending_reqs = Reqs1}.

